app/src/main/java/edu/usf/sas/pal/muser/firebase/models/CurrentSongInfo.kt

```diff
@@ -0,0 +1,32 @@
+package edu.usf.sas.pal.muser.firebase.models
+
+
+/**
+ * A model class that holds the information about the song that is currently being played by the
+ * user.
+ */
+data class CurrentSongInfo
+
+/**
+ * [id] - The id of the track.
+ * [name] - Name of the track.
+ * [artistName] - Name of the Artist who composed the track.
+ * [albumName] - Name of the album that the current track belongs to.
+ * [duration] - The length of the track.
+ * [dateAdded] - The date on which the track was added.
+ * [year] - The year during which the track was released.
+ * [playCount] - The number of times the user played this track.
+ * [sampleRateLabel] - The sample rate of the song.
+ * [formatLabel] - The format of the song. (mp3 or Flac etc.)
+ */
+(val id: Long,
```

Comment:
Are all of these values guaranteed to be populated? If not, we need to mark some (all?) of them optional instead of required. With Kotlin we shouldn't be passing `null` values into the constructor.

Reply:
I can do that. Since the song class is populated entirely every time the user plays a song, I thought we could keep all the field as required. However, I do see your point. If I edit tags in the mp3, then the Song object will have null values.

Reply:
I think it's better if we keep the ID values and the name of the song to be mandatory because I don't think it makes much sense to have a song without a name?

Reply:
If the ID is generated by MUSER and always exists, then yes, it should be required.

I agree that it doesn't make sense to have a song without a name, but we have to consider all possible values that MUSER will encounter at runtime. It looks like the [ID3](https://en.wikipedia.org/wiki/ID3) spec says that tags without values are empty strings (i.e., there is no `null` vs. empty differentiation), and the title could certainly be missing from a tag. So maybe it makes sense to require name, and just populate it with an empty string if it doesn't exist. 

This makes me wonder, though - should we follow the same process for the remaining string fields?

Please take a look at how Android and the content provider handles empty fields - if they differentiate between `null` and empty string we should as well. If not, then it probably makes sense to store empty strings. In other words, let's follow the behavior of Android.

Reply:
> If the ID is generated by MUSER and always exists, then yes, it should be required.

To further narrow the `_ID` topic down, I think the ID is generated by the `ContentProvider` while it scans the storage and stores the data in a relational way. If we were to store the DeviceID or anything that can uniquely identify a device along with the locally generated ID and use them as one primary key, I think we'll able to uniquely identify a "Song" for a device from server-side?

Reply:
> Please take a look at how Android and the content provider handles empty tag fields - if they differentiate between null and empty string we should as well. If not, then it probably makes sense to store empty strings. In other words, let's follow the behavior of Android.

I tried logging what the `ContentProvider` would return for empty tags and here's what I got:-
`2020-08-07 14:06:31.698 27838-27838/edu.usf.sas.pal.muser D/PlayerFragment: trackInfoChanged: <Unknown>`

I'm assuming there won't be a null object as `ContentProvider` returns `<Unknown>` by default.

Reply:
> If we were to store the DeviceID or anything that can uniquely identify a device along with the locally generated ID and use them as one primary key, I think we'll able to uniquely identify a "Song" for a device from server-side?

Yes, makes sense. We'll generate a DeviceID separately (and can talk about that later). So let's store the song ID.

Reply:
>I tried logging what the ContentProvider would return for empty tags and here's what I got:-
2020-08-07 14:06:31.698 27838-27838/edu.usf.sas.pal.muser D/PlayerFragment: trackInfoChanged: <Unknown>

Hmmm, ok. I guess let's make the following required, since they will almost always be populated (and if they're not we'll just set them to an empty string):

```kotlin
 val name: String,
 val artistName: String,
 val albumName: String,
```

All other strings should be optional.

Reply:
Sounds great!